feat: generate `Parse.Object.objectId` automatically when `allowCustomObjectId` is enabled and no `objectId` is passed

musthafa1996 opened the pull request:
### New Pull Request Checklist
<!--
    Check every following box [x] before submitting your PR.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Platform!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/Parse-SDK-JS/issues?q=is%3Aissue).

### Issue Description
Enabling `allowCustomObjectId` forces all objects to have custom `objectId`s instead of auto-generating when no `objectId` is passed, or when `objectId` is **null** or **undefined**. 

Related issue: https://github.com/parse-community/Parse-SDK-JS/issues/1402

### Approach
When `allowCustomObjectId` is enabled:
- Check if `objectId` has a custom value.
- If `objectId` has a value and is an empty string, throw 'objectId must not be empty' error.
- If `objectId` doesn't have a value, or is **null** or **undefined**, generate `objectId` automatically as if `allowCustomObjectId` were disabled.

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete TODOs that do not apply to this PR.
-->

- [x ] Add tests
- [ ] Add entry to changelog
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)

musthafa1996 committed:
feat: Update conditions for `MISSING_OBJECT_ID` error when allowCustomObjectId is enabled

musthafa1996 committed:
Update `MISSING_OBJECT_ID` error message

musthafa1996 committed:
Update tests

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2063,18 +2063,32 @@ describe('Parse Object', () => {
+    const object1 = new Parse.Object('TestObject');
     try {
-      await object.save();
+      await object1.save();
+    } catch (error) {
+      fail();
+    }
```
```suggestion
    await object1.save();
```

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2092,17 +2106,23 @@ describe('Parse Object', () => {
+    obj1.id = '';
     const obj2 = new TestObject({ foo: 'baz' });
+    obj2.id = '';
     try {
       await Parse.Object.saveAll([obj1, obj2]);
       fail();
     } catch (error) {
```
```suggestion
    await expectAsync(Parse.Object.saveAll([obj1, obj2])).toBeRejectedWith(new ParseError(ParseError.MISSING_OBJECT_ID, 'objectId must not be empty'))
```

dblythy reviewed (changes requested):
Looks mostly good, just some minor refactoring

musthafa1996 committed:
Update integration/test/ParseObjectTest.js

Co-authored-by: dblythy <daniel-blyth@live.com.au>

musthafa1996 committed:
Update integration/test/ParseObjectTest.js

Co-authored-by: dblythy <daniel-blyth@live.com.au>

musthafa1996 commented:
Thanks for reviewing @dblythy! I have refactored the tests with your suggestions. Can you check again please?

musthafa1996 commented:
Hi @dblythy! Were you able to have a look at the recent changes pushed to the PR?

dblythy commented:
You might need to run `npm run lint` and fix syntax errors. I couldn't suggest changes to be syntactically correct as I couldn't provide suggestions across deleted lines.

musthafa1996 commented:
> You might need to run `npm run lint` and fix syntax errors. I couldn't suggest changes to be syntactically correct as I couldn't provide suggestions across deleted lines.

Thanks. I’ll run and post the results.

musthafa1996 committed:
Fix tests

musthafa1996 commented:
@dblythy, my bad. I have just fixed the syntax error and updated the PR. Please have a look now.

The following is the result of `npm run lint`:
![Screenshot 2022-09-06 at 3 55 43 PM](https://user-images.githubusercontent.com/22429514/188612077-9fdf1d17-ae28-44ce-83c3-45bd7eca9874.png)

The following is the result of `npm run test`:
![Screenshot 2022-09-06 at 3 56 39 PM](https://user-images.githubusercontent.com/22429514/188612261-2696c278-23d2-4296-bad4-ecc5584ba5ee.png)
![Screenshot 2022-09-06 at 3 56 59 PM](https://user-images.githubusercontent.com/22429514/188612344-90a86b62-49b4-4f70-9e1d-efc259fa1c25.png)

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2092,17 +2102,27 @@ describe('Parse Object', () => {
+    );
+
     try {
       await Parse.Object.saveAll([obj1, obj2]);
       fail();
     } catch (error) {
-      expect(error.message).toBe('objectId must not be empty, null or undefined');
+      expect(error.message).toBe('objectId must not be empty');
     }
```
```suggestion
```

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2063,18 +2063,28 @@ describe('Parse Object', () => {
+    object2.id = '';
     try {
-      await object.save();
+      await object2.save();
       fail();
     } catch (error) {
-      expect(error.message).toBe('objectId must not be empty, null or undefined');
+      expect(error.message).toBe('objectId must not be empty');
     }
```
```suggestion
    await expectAsync(object2.save()).toBeRejectedWith(
      new Parse.Error(Parse.Error.MISSING_OBJECT_ID, 'objectId must not be empty')
    );

```

dblythy reviewed (changes requested):
Sorry, I just worked out how to properly suggest 😂. 

Almost there, @mtrezza can you allow the tests to run?

musthafa1996 committed:
Fix tests for custom object ID

musthafa1996 commented:
I had been careful enough this time while refactoring. Can you check yet again, @dblythy ? :)

musthafa1996 commented:
@mtrezza, @dblythy, have you checked the latest changes?

dblythy commented:
I will review again within 24 hrs 😊

musthafa1996 commented:
Thank you 😊

dblythy commented:
Can you check the integration check? `npm run integration`

mtrezza committed:
Merge branch 'alpha' into allow-custom-id

musthafa1996 committed:
Remove unnecessary id existence check from ParseObject in case allowCustomObjectId is true

musthafa1996 committed:
Fix tests for allowCustomObjectId

musthafa1996 commented:
> Can you check the integration check? `npm run integration`

Hi @dblythy and @mtrezza! I did get some time to look into this PR today. I have pushed the following updates:

- Remove unnecessary id existence check from `ParseObject` in case allowCustomObjectId is true.
- Fix tests for allowCustomObjectId.

Turns out that we don't need the following checks in `ParseObject` as they are already validated from ParseServer.

![Screenshot 2022-09-10 at 2 09 40 AM](https://user-images.githubusercontent.com/22429514/189439291-cbd0cb1e-9a97-457d-8670-709b62f1887e.png)

![Screenshot 2022-09-10 at 2 09 55 AM](https://user-images.githubusercontent.com/22429514/189439321-4d7c8b8b-15c0-4325-b747-07dcefc42913.png)

All the related tests seem to work fine now.

musthafa1996 commented:
There is one thing to note here. The error received from ParseServer says 'objectId must not be empty, null or undefined'. Yes, object saving fails when the `objectId` is **empty** or **null**. However, saving succeeds when `objectId` is **undefined** with automatic objectId generated. 

We probably need to update the error message on ParseServer. Please advice.

musthafa1996 committed:
Merge branch 'alpha' into allow-custom-id

dblythy commented:
> Turns out that we don't need the following checks in ParseObject as they are already validated from ParseServer

Even though the checks happen on Parse Server, the client side validation prevents unnecessary computation / network calls. There's no downside to have them run on the client side as well

musthafa1996 commented:
Okay. Let me add them back in that case.

musthafa1996 committed:
Add custom objectId validation back to ParseObject

musthafa1996 commented:
@dblythy, @mtrezza, I have added the validation back. All the test cases are also passed.

musthafa1996 committed:
Fix ParseObject-test.js

mtrezza commented:
Could you take a look at the failing CI? The lock file version needs to be 1, then the "ci build" will likely run.

musthafa1996 committed:
Fix package-lock.json

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2087,22 +2101,35 @@ describe('Parse Object', () => {
   });
 
-  it('allowCustomObjectId saveAll', async () => {
+  it('allowCustomObjectId saveAll', async done => {
```
no need for done
```suggestion
  it('allowCustomObjectId saveAll', async () => {
```

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2114,5 +2141,7 @@ describe('Parse Object', () => {
       expect([customId1, customId2].includes(result.id));
     });
     Parse.allowCustomObjectId = false;
+
+    done();
```
```suggestion
```

musthafa1996 commented:
@mtrezza, I have fixed the **packages-lock.json** file. Can you run the build again?

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2087,22 +2101,35 @@ describe('Parse Object', () => {
     const afterSave = await query.get(customId);
     expect(afterSave.get('foo')).toBe('baz');
     Parse.allowCustomObjectId = false;
+
+    done();
```
```suggestion
```

dblythy commented on integration/test/ParseObjectTest.js:
```diff
@@ -2060,21 +2060,35 @@ describe('Parse Object', () => {
   });
 
-  it('allowCustomObjectId', async () => {
+  it('allowCustomObjectId', async done => {
```
```suggestion
  it('allowCustomObjectId', async () => {
```

dblythy reviewed (changes requested):
Almost there!

musthafa1996 committed:
Cleanup ParseObjectTest.js

musthafa1996 commented:
@dblythy, pushed the **ParseObjectTest.js** updates. Can you check again?

musthafa1996 commented:
@dblythy, can you do a review?

mtrezza committed:
Merge branch 'alpha' into allow-custom-id

dblythy commented:
It's mostly good - just need to remove the `done()` from the tests

musthafa1996 committed:
Merge branch 'allow-custom-id' of github.com:musthafa1996/Parse-SDK-JS into allow-custom-id

musthafa1996 commented:
@dblythy, the test starts failing when I remove `done()` from `allowCustomObjectId` test in **ParseObject-test.js**. Please check the screenshots below and share your thoughts.

With `done()`:
![Screenshot 2022-09-16 at 9 16 24 PM](https://user-images.githubusercontent.com/22429514/190678628-b68aeb2c-943b-4b8b-a85b-ae4b6ede17e8.png)

Without `done()`:
![Screenshot 2022-09-16 at 9 17 58 PM](https://user-images.githubusercontent.com/22429514/190678912-a037d8c1-49f0-40cc-945f-e736f0a51a28.png)

mtrezza committed:
Merge branch 'alpha' into allow-custom-id

musthafa1996 commented:
@dblythy any idea why?

dblythy commented:
when you remove `done()` you also must remove `async done => {`

So, remove `done()` but also change the function to `async () => {`

musthafa1996 commented:
Thanks :) 
Let me try that. I'm new to testing by the way :D

dblythy commented:
No worries! It should work if you make the changes as per my [previous review](https://github.com/parse-community/Parse-SDK-JS/pull/1540#pullrequestreview-1103567948). Feel free to let us know if you need any more help!

musthafa1996 commented:
@dblythy, this is what I get after removing `done()`. Lint doesn't identify `fail()` method at lines 3822 and 3828,
![Screenshot 2022-09-19 at 12 23 32 PM](https://user-images.githubusercontent.com/22429514/190963744-956bcc3e-454c-432c-b164-ce107b44f08e.png)

musthafa1996 committed:
Fix ParseObject-test.js

musthafa1996 committed:
Merge branch 'allow-custom-id' of github.com:musthafa1996/Parse-SDK-JS into allow-custom-id

musthafa1996 commented:
Thanks @dblythy. I have resolved the tests. Please review.

mtrezza committed:
Merge branch 'alpha' into allow-custom-id

musthafa1996 commented:
What's pending for this PR to be merged now @dblythy, @mtrezza?

mtrezza commented on integration/test/ParseObjectTest.js:
```diff
@@ -2063,18 +2063,32 @@ describe('Parse Object', () => {
     await reconfigureServer({ allowCustomObjectId: true });
     Parse.allowCustomObjectId = true;
+
+    // Try to save without passing objectId
```
If a test title doesn't tell exactly what it tests and it needs comments to explain, that's usually a sign that it should be split into individual tests. Combining multiple test scenarios in a single test increases code complexity and makes it more difficult to add changes in the future.

You could add a new `describe` layer for `allowCustomObjectId` and the add tests like:
```
describe('allowCustomObjectId', () => {
  it('fails to save with empty objectId', async () => {
```

Could you split this into individual tests?

musthafa1996 replied:
Yes, sure. It was like that previously.

musthafa1996 replied:
It was like that previously.

musthafa1996 replied:
Hi @mtrezza! I have split the tests for allowCustomObjectId. Please review again and mark this request as resolved if all is good.

mtrezza commented on src/ParseObject.js:
```diff
@@ -2419,6 +2413,13 @@ const DefaultController = {
+              if (allowCustomObjectId && Object.prototype.hasOwnProperty.call(el, 'id') && !el.id) {
+                throw new ParseError(
+                  ParseError.MISSING_OBJECT_ID,
+                  'objectId must not be empty or null'
```
This has changed from `'objectId must not be empty, null or undefined'` to ` 'objectId must not be empty or null'`. Why is `undefined` not allowed?

musthafa1996 committed:
Split ParseOnject allowCustomObjectId integration tests

musthafa1996 committed:
Merge branch 'allow-custom-id' of github.com:musthafa1996/Parse-SDK-JS into allow-custom-id

mtrezza committed:
Merge branch 'alpha' into allow-custom-id

mtrezza commented:
Waiting for CI to pass

musthafa1996 commented:
CI has been completed successfully.

mtrezza reviewed (approved):
Looks good!

mtrezza commented:
@dblythy Is this ready for merge also from your side?

musthafa1996 commented:
@mtrezza, is there any actions required from me to get this merged?

musthafa1996 committed:
Merge branch 'alpha' into allow-custom-id

parseplatformorg commented:
🎉 This change has been released in version [3.5.0-alpha.4](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.0-alpha.4)

musthafa1996 commented:
Thanks for merging @mtrezza and @dblythy :)

dblythy commented:
Thanks for your contribution!! 🔥

musthafa1996 commented:
Hi @mtrezza! When can we expect version 3.5.0 to be released?

mtrezza commented:
The release cycle is monthly (around the beginning of every month): alpha -> beta -> release
This feature will be in beta in Okt. and in stable release in Nov.

musthafa1996 commented:
Got it. Thanks for letting me know @mtrezza 😊

parseplatformorg commented:
🎉 This change has been released in version [3.5.0-beta.1](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.0-beta.1)

parseplatformorg commented:
🎉 This change has been released in version [3.5.0](https://github.com/parse-community/Parse-SDK-JS/releases/tag/3.5.0)
